### ISSUE TITLE
Email notifier doesn't seem to work

### ISSUE THREAD

**Issue by gshilo** [linkedin/Burrow#165]
Hello

I'm trying to test burrow. I installed and configured it on my test computer (along with Kafka and a mail server).
The only notofier I need is the Email notifier, so I removed all other notifiers from the config file. I also do not have Storm.
All I want is to receive email alerts.
So my config file looks like this:
[general]
logdir=/home/kafka/Burrow-master/log
logconfig=/home/kafka/Burrow-master/config/logging.cfg
pidfile=burrow.pid
client-id=burrow-lagchecker
group-blacklist=^(console-consumer-|python-kafka-consumer-).*$
#group-whitelist=^(my-important-consumer).*$

[zookeeper]
hostname=hadoop.localdomain
port=2181
timeout=15
lock-path=/burrow/notifier

[kafka "guy1"]
broker=hadoop.localdomain
broker-port=9092
zookeeper=hadoop.localdomain
zookeeper-port=2181
zookeeper-path=/kafka-cluster
offsets-topic=__consumer_offsets

[tickers]
broker-offsets=60

[lagcheck]
intervals=10
expire-group=604800

[httpserver]
server=on
port=8000

[smtp]
server=10.0.2.2
port=25
username=guy
password=a123456
auth-type=plain
from=burrow-noreply@example.com
template=/home/kafka/Burrow-master/config/default-email.tmpl

[emailnotifier "ni4ni@hotmail.com"]
group=guy1,critical-consumer-group
group=guy1,other-consumer-group
group=guy1,test
interval=30
enable=true

[notify]
interval=10

I then wrote a producer and a consumer that halts after few minutes to create a lag.
About a minute after the consumer hangs, the status changes from "OK" to "Error" and I can see it if I look at the status in the web server.
However, no mail notification is sent. There are also no errors in burrow log.
I expect that if something is wrong with the email setup I will get an error or a message in the log file, but it seems like Email notification just doesn't work.

This is how the log looks like:

2016-12-15 23:05:37 [INFO] Starting Zookeeper client
2016-12-15 23:05:37 [INFO] Starting Offsets Storage module
2016-12-15 23:05:37 [INFO] Starting HTTP server
2016-12-15 23:05:37 [INFO] Starting Zookeeper client for cluster guy1
2016-12-15 23:05:37 [INFO] Starting Kafka client for cluster guy1
2016-12-15 23:05:37 [INFO] Starting consumers for 50 partitions of __consumer_offsets in cluster guy1
2016-12-15 23:06:03 [INFO] Start email notify
2016-12-15 23:06:03 [INFO] Acquired Zookeeper notify lock
2016-12-15 23:11:03 [INFO] Start evaluating consumer group logstash in cluster guy1
2016-12-15 23:11:03 [INFO] Start evaluating consumer group test in cluster guy1

What is wrong with the Email notifier ? or with my setup ?

Thank you

Guy

**Comment by asifhj** [comment]
Same with me also! Any pointer to look into ?

**Comment by dgqypl** [comment]
@gshilo @asifhj When burrow started up a while, is there any log in logfile like this:
[ERROR] Failed to send email message:unencrypted connection

**Comment by toddpalino** [comment]
In general, yes, you should see an error if the email fails to send. If you set your log level to debug, you should at least see a "send email" message at that level when the email tries to send.

**Comment by toddpalino** [comment]
Upon second look while looking at #145, this is because not all of the consumer groups that you have specified for email are in a bad state. This is incorrect behavior - the email should get sent if any group is in the bad state, not all. However, with the way the notifier center works right now, groups are evaluated in a way that makes it hard to collect this information.

What probably needs to happen here is that each notifier needs to collect information on each group, and then fire off the notification on an interval. This may require a new timer, but it will definitely require more investigation into the code before I can come up with the change required.

**Comment by gshilo** [comment]
Thank you for your response.
This was happening when I started Burrow and only then started Kafka. When I did it in a reverse order it worked.

**Comment by JarenGlover** [comment]
@toddpalino So I am trying to get around the whole "if all consumers in a `emailnotifer` are not in a "bad" state then no alarm" issue by creating multiple `emailnotifer` sections. 

However, it appears the config can't handle multiple `emailnotifer` sections. example config below. If this is the case than is the expectation I will need to run multiple instances of burrow to get around this problem? 

```
[emailnotifier "ops@example.com"]
group=production,a
interval=60
enable=true
threshold=2

[emailnotifier "ops@example.com"]
group=production,b
interval=60
enable=true
threshold=2

[emailnotifier "ops@example.com"]
group=production,c
interval=60
enable=true
threshold=2
```

**Comment by toddpalino** [comment]
Multiple email notifiers should work properly. What is the behavior you're seeing when you configure this way?

**Comment by JarenGlover** [comment]
@toddpalino no emails being sent. log is at `trace` level however, not showing any emails being sent. when update my config and only remove the extra two notifiers everything works as expected.

**Comment by toddpalino** [comment]
Oh, wait, I see the problem. The issue is that the email address in the config section header is used as a key. This means that you cannot configure multiple email notifiers with the same email address. So the behavior you're seeing is expected, but wrong.

We'll need a PR to fix this that moves the email address to an explicit config in the section, and the key in the header can then be freeform and should be required to be unique. To maintain current behavior, we should check in config.go that if the email address config is missing, and the key is a valid email address, we'll use that instead.

**Comment by JarenGlover** [comment]
@toddpalino blah. i meant to check if that could be the problem (email being unique key). 

sorry. thanks for quick reply.

**Comment by plutoid** [comment]
A little question for such issue, how linkedin set the email notification on their production environment? do they use this?

**Comment by jeffmontagna** [comment]
I created a pull request that addresses using the email address as a unique key.  https://github.com/linkedin/Burrow/pull/245

**Comment by toddpalino** [comment]
#245 is the solution to this. I'm going to pull it into the rewrite and make sure we don't have this problem with any module.

**Comment by toddpalino** [comment]
Resolved by 1.0 rewrite
